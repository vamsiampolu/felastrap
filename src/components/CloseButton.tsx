import type { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { StyleProps, Theme } from "./common";

const close = () => {
  const borderRadius = {
    borderTopLeftRadius: "0.375rem",
    borderTopRightRadius: "0.375rem",
    borderBottomRightRadius: "0.375rem",
    borderBottomLeftRadius: "0.375rem",
  };

  return {
    ...borderRadius,
    borderColor: "rgb(0,0,0)",
    backgroundColor: "transparent",
    backgroundImage: `url(${closeIcon})`,
    backgroundPositionX: "center",
    backgroundOrigin: "padding-box",
    backgroundClip: "border-box",
    backgroundPositionY: "1em",
    backgroundRepeat: "no-repeat",
    backgroundSize: "auto",
    boxSizing: "content-box",
    border: 0,
    color: "#000",
    height: "1em",
    opacity: 0.5,
    paddingTop: "0.25em",
    paddingBottom: "0.25em",
    paddingLeft: "0.25em",
    paddingRight: "0.25em",
    width: "1em",
    ":focus": {
      boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
      opacity: 1,
      outline: 0,
    },
    ":disabled": {
      opacity: 0.25,
      pointerEvents: "none",
      userSelect: "none",
    },
  };
};

const alertRule = ({ alert }: StyleProps<CloseButtonProps>) => {
  return alert
    ? {
        paddingTop: "1.25rem",
        paddingBottom: "1.25rem",
        paddingLeft: "1rem",
        paddingRight: "1rem",
        position: "absolute",
        right: 0,
        top: 0,
        zIndex: 2,
      }
    : {};
};

const modalRule = ({ modal }: StyleProps<CloseButtonProps>) =>
  modal
    ? {
        padding: "0.5rem 0.5rem",
        margin: "-0.5rem -0.5rem -0.5rem auto",
      }
    : {};

export type CloseButtonProps = ComponentPropsWithoutRef<"button"> &
  Partial<{ alert?: boolean; modal?: boolean }>;

export const CloseButton = ({
  alert = false,
  modal = false,
  className = "",
  ...other
}: CloseButtonProps) => {
  const { css } = useFela<Theme, CloseButtonProps>({ alert, modal });
  const cx = `${css([
    close,
    alertRule,
    modalRule,
  ])} close-button ${className}`.trim();
  return (
    <button
      data-testid="close-button"
      className={cx}
      aria-label="Close"
      {...other}
    />
  );
};
