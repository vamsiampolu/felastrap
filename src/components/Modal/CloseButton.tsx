import { ComponentPropsWithoutRef } from "react";
import { useFela } from "react-fela";
import close from "/close-button.svg";

const rule = () => ({
  marginTop: "-2px",
  appearance: "none",
  textAlign: "end",
  cursor: "pointer",
  border: 0,
  position: "absolute",
  right: "2rem",
  top: "1rem",
  padding: "1rem",
  backgroundColor: "transparent",
  backgroundImage: `url(${close})`,
  backgroundRepeat: "no-repeat",
  backgroundSize: "16px",
  backgroundPosition: "center",
});

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "type" | "aria-label"
>;

export const CloseButton = ({
  onClick,
  className = "",
  ...props
}: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} bs-close ${className}`.trim();
  return (
    <button
      data-testid="modal-close-button"
      className={cx}
      onClick={onClick}
      type="button"
      aria-label="Close"
      {...props}
    ></button>
  );
};
