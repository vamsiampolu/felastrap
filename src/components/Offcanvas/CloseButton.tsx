import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { ComponentPropsWithoutRef } from "react";

const borderRadius = {
  borderTopLeftRadius: "0.375rem",
  borderTopRightRadius: "0.375rem",
  borderBottomRightRadius: "0.375rem",
  borderBottomLeftRadius: "0.375rem",
};

const rule = () => ({
  ...borderRadius,
  borderColor: "rgb(0,0,0)",
  backgroundColor: "transparent",
  backgroundImage: `url(${closeIcon})`,
  backgroundPositionX: "center",
  backgroundOrigin: "padding-box",
  backgroundClip: "border-box",
  backgroundPositionY: "1em",
  backgroundRepeat: "no-repeat",
  backgroundSize: "auto",
  boxSizing: "content-box",
  border: 0,
  color: "#000",
  height: "1em",
  opacity: 0.5,
  paddingTop: "10px",
  paddingBottom: "10px",
  paddingLeft: "0.25em",
  paddingRight: "0.25em",
  width: "1em",
  ":focus": {
    boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
    opacity: 1,
    outline: 0,
  },
  ":disabled": {
    opacity: 0.25,
    pointerEvents: "none",
    userSelect: "none",
  },
});

const offcanvasRule = () => ({
  padding: "calc(1rem * 0.5) calc(1rem * 0.5)",
  marginTop: "calc(-0.5 * 1rem)",
  marginRight: "calc(-0.5 * 1rem)",
  marginBottom: "calc(-0.5 * 1rem)",
});

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "aria-label" | "type" | "children"
>;

export const CloseButton = ({ className = "", ...props }: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css([
    rule,
    offcanvasRule,
  ])} offcanvas-btn-close ${className}`.trim();
  return (
    <button
      data-testid="offcanvas-close-button"
      type="button"
      className={cx}
      aria-label="Close"
      {...props}
    />
  );
};
