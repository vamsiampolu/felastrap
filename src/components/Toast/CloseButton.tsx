import { useFela } from "react-fela";
import closeIcon from "/close-button.svg";
import { ComponentPropsWithoutRef } from "react";

const borderRadius = {
  borderTopLeftRadius: "0.375rem",
  borderTopRightRadius: "0.375rem",
  borderBottomRightRadius: "0.375rem",
  borderBottomLeftRadius: "0.375rem",
};

const rule = () => {
  return {
    ...borderRadius,
    marginRight: "calc(-0.5 * 0.75rem)",
    marginLeft: "0.75rem",
    paddingTop: "10px !important",
    paddingBottom: "10px !important",
    borderColor: "rgb(0,0,0)",
    backgroundColor: "transparent",
    backgroundImage: `url(${closeIcon})`,
    backgroundPositionX: "center",
    backgroundOrigin: "padding-box",
    backgroundClip: "border-box",
    backgroundPositionY: "1em",
    backgroundRepeat: "no-repeat",
    backgroundSize: "auto",
    boxSizing: "content-box",
    border: 0,
    color: "#000",
    height: "1em",
    opacity: 0.5,
    width: "1em",
    ":focus": {
      boxShadow: "0 0 0 0.25rem rgba(13, 110, 253, 0.25)",
      opacity: 1,
      outline: 0,
    },
    ":disabled": {
      opacity: 0.25,
      pointerEvents: "none",
      userSelect: "none",
    },
  };
};

export type CloseButtonProps = Omit<
  ComponentPropsWithoutRef<"button">,
  "aria-label" | "type" | "children"
>;
export const CloseButton = ({ className = "", ...props }: CloseButtonProps) => {
  const { css } = useFela();
  const cx = `${css(rule)} toast-btn-close ${className}`;
  return (
    <button
      data-testid="toast-close-button"
      type="button"
      className={cx}
      aria-label="Close"
      {...props}
    />
  );
};
