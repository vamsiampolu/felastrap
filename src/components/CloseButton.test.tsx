import { IRenderer } from "fela";
import {
  Wrapper,
  createRenderer,
  cssReboot,
  createSnapshot,
} from "../test-helpers";
import { CloseButton } from "./CloseButton";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("components/CloseButton", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a close button", async () => {
    const component = <CloseButton />;
    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;

    render(wrappedComponent);

    expect(screen.getByTestId("close-button")).toBeInTheDocument();
    expect(screen.queryByTestId("close-button")).toHaveClass("close-button");

    expect(screen.queryByTestId("close-button")).toHaveAccessibleName("Close");
    expect(screen.queryByTestId("close-button")?.tagName).toEqual("BUTTON");

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a close button with a className", async () => {
    const component = <CloseButton className="close" />;
    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;

    render(wrappedComponent);

    expect(screen.getByTestId("close-button")).toBeInTheDocument();
    expect(screen.queryByTestId("close-button")).toHaveClass("close");

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a close button for an alert", async () => {
    const component = <CloseButton alert />;
    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;

    render(wrappedComponent);
    expect(screen.getByTestId("close-button")).toBeInTheDocument();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a close button for a modal", async () => {
    const component = <CloseButton modal />;
    const wrappedComponent = <Wrapper renderer={renderer}>{component}</Wrapper>;

    render(wrappedComponent);
    expect(screen.getByTestId("close-button")).toBeInTheDocument();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should invoke onClick prop when clicked", async () => {
    const onClick = jest.fn();
    const user = userEvent.setup();

    const component = (
      <Wrapper renderer={renderer}>
        <CloseButton onClick={onClick} />
      </Wrapper>
    );
    render(component);

    const btn = screen.queryByTestId("close-button");
    expect(btn).toBeInTheDocument();

    await user.click(btn!);
    expect(onClick).toHaveBeenCalled();
  });
});
