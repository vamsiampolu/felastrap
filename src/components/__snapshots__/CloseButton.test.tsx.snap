// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CloseButton should render a close button 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  background-color: transparent;
}
.f {
  background-image: url(test-file-stub);
}
.g {
  background-position-x: center;
}
.h {
  background-origin: padding-box;
}
.i {
  background-clip: border-box;
}
.j {
  background-position-y: 1em;
}
.k {
  background-repeat: no-repeat;
}
.l {
  background-size: auto;
}
.m {
  box-sizing: content-box;
}
.n {
  color: #000;
}
.o {
  height: 1em;
}
.p {
  opacity: 0.5;
}
.q {
  padding-top: 0.25em;
}
.r {
  padding-bottom: 0.25em;
}
.s {
  padding-left: 0.25em;
}
.t {
  padding-right: 0.25em;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  border-top-width: 0;
}
.ae {
  border-right-width: 0;
}
.af {
  border-bottom-width: 0;
}
.ag {
  border-left-width: 0;
}
.ah {
  border-top-color: rgb(0, 0, 0);
}
.ai {
  border-right-color: rgb(0, 0, 0);
}
.aj {
  border-bottom-color: rgb(0, 0, 0);
}
.ak {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="close-button"
  class="a ab ac ae af ag ah ai aj ak b c d e f g h i j k l m n o p q r s t u v w x y z close-button"
  aria-label="Close"
></button
>;
"
`;

exports[`components/CloseButton should render a close button for a modal 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  background-color: transparent;
}
.f {
  background-image: url(test-file-stub);
}
.g {
  background-position-x: center;
}
.h {
  background-origin: padding-box;
}
.i {
  background-clip: border-box;
}
.j {
  background-position-y: 1em;
}
.k {
  background-repeat: no-repeat;
}
.l {
  background-size: auto;
}
.m {
  box-sizing: content-box;
}
.n {
  color: #000;
}
.o {
  height: 1em;
}
.p {
  opacity: 0.5;
}
.q {
  padding-top: 0.25em;
}
.r {
  padding-bottom: 0.25em;
}
.s {
  padding-left: 0.25em;
}
.t {
  padding-right: 0.25em;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  margin-top: -0.5rem;
}
.ae {
  margin-right: -0.5rem;
}
.af {
  margin-bottom: -0.5rem;
}
.ag {
  margin-left: auto;
}
.ah {
  border-top-width: 0;
}
.ai {
  border-right-width: 0;
}
.aj {
  border-bottom-width: 0;
}
.ak {
  border-left-width: 0;
}
.al {
  border-top-color: rgb(0, 0, 0);
}
.am {
  border-right-color: rgb(0, 0, 0);
}
.an {
  border-bottom-color: rgb(0, 0, 0);
}
.ao {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="close-button"
  class="a ab ac ae af ag ah ai aj ak al am an ao b c d e f g h i j k l m n o p q r s t u v w x y z close-button"
  aria-label="Close"
></button
>;
"
`;

exports[`components/CloseButton should render a close button for an alert 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  background-color: transparent;
}
.f {
  background-image: url(test-file-stub);
}
.g {
  background-position-x: center;
}
.h {
  background-origin: padding-box;
}
.i {
  background-clip: border-box;
}
.j {
  background-position-y: 1em;
}
.k {
  background-repeat: no-repeat;
}
.l {
  background-size: auto;
}
.m {
  box-sizing: content-box;
}
.n {
  color: #000;
}
.o {
  height: 1em;
}
.p {
  opacity: 0.5;
}
.q {
  padding-top: 1.25rem;
}
.r {
  padding-bottom: 1.25rem;
}
.s {
  padding-left: 1rem;
}
.t {
  padding-right: 1rem;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  position: absolute;
}
.ae {
  right: 0;
}
.af {
  top: 0;
}
.ag {
  z-index: 2;
}
.ah {
  border-top-width: 0;
}
.ai {
  border-right-width: 0;
}
.aj {
  border-bottom-width: 0;
}
.ak {
  border-left-width: 0;
}
.al {
  border-top-color: rgb(0, 0, 0);
}
.am {
  border-right-color: rgb(0, 0, 0);
}
.an {
  border-bottom-color: rgb(0, 0, 0);
}
.ao {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="close-button"
  class="a ab ac ae af ag ah ai aj ak al am an ao b c d e f g h i j k l m n o p q r s t u v w x y z close-button"
  aria-label="Close"
></button
>;
"
`;

exports[`components/CloseButton should render a close button with a className 1`] = `
".a {
  border-top-left-radius: 0.375rem;
}
.b {
  border-top-right-radius: 0.375rem;
}
.c {
  border-bottom-right-radius: 0.375rem;
}
.d {
  border-bottom-left-radius: 0.375rem;
}
.e {
  background-color: transparent;
}
.f {
  background-image: url(test-file-stub);
}
.g {
  background-position-x: center;
}
.h {
  background-origin: padding-box;
}
.i {
  background-clip: border-box;
}
.j {
  background-position-y: 1em;
}
.k {
  background-repeat: no-repeat;
}
.l {
  background-size: auto;
}
.m {
  box-sizing: content-box;
}
.n {
  color: #000;
}
.o {
  height: 1em;
}
.p {
  opacity: 0.5;
}
.q {
  padding-top: 0.25em;
}
.r {
  padding-bottom: 0.25em;
}
.s {
  padding-left: 0.25em;
}
.t {
  padding-right: 0.25em;
}
.u {
  width: 1em;
}
.y:disabled {
  opacity: 0.25;
}
.z:disabled {
  pointer-events: none;
}
.ab:disabled {
  user-select: none;
}
.ac {
  border-top-width: 0;
}
.ae {
  border-right-width: 0;
}
.af {
  border-bottom-width: 0;
}
.ag {
  border-left-width: 0;
}
.ah {
  border-top-color: rgb(0, 0, 0);
}
.ai {
  border-right-color: rgb(0, 0, 0);
}
.aj {
  border-bottom-color: rgb(0, 0, 0);
}
.ak {
  border-left-color: rgb(0, 0, 0);
}
.v:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.w:focus {
  opacity: 1;
}
.x:focus {
  outline-width: 0;
}


<button
  data-testid="close-button"
  class="a ab ac ae af ag ah ai aj ak b c d e f g h i j k l m n o p q r s t u v w x y z close-button close"
  aria-label="Close"
></button
>;
"
`;
